_posts/2020-09-06-how-the-system-tests-works.md

```diff
@@ -0,0 +1,501 @@
+`Apache Kafka Clients` for the external communication, Kubernetes client, `Constants.java` and `Environment.java`.
+Main idea is to make the system tests and its resources easily modifiable and writable in the fluent way. Before, we dive into
+our `Resources` we need to understand the lifecycle of system tests. 
+
+## Lifecycle of tests
+
+Like any other test life cycle, we follow the [SEVT](http://xunitpatterns.com/Four%20Phase%20Test.html) convention. 
+[SEVT](http://xunitpatterns.com/Four%20Phase%20Test.html) stands for Setup, Exercise, Verify and Teardown.
+
+### Setup
+
+In `Setup phase` we perform following actions:
+
+* Create namespace(s)
+* Deploy the Strimzi Cluster operator
+* Deploy the Kafka cluster and/or other components (optional)
+
+The reason why the last point is optional, is because we have some test cases where we might want to have a different kafka 
+configuration for each test scenario, so creation of the Kafka cluster and other resources is done in the test phase.
+
+We create resources in Kubernetes cluster via classes in `resources` package, which allows you to deploy all components and, 
+if needed, change theirs default configuration using a builder.
+Currently, we have three stacks, which are stored in `ResourceManager` singleton instance — one for all test class resources, 
+one for test method resources and one auxiliary, which always points to class or method stack.
+You are allowed to create resources during execution of any test method. 
+Resource lifecycle implementation handles insertion of the resource on top of the stack and its deletion at the end of the test method/class (teardown phase).
+
+`ResourceManager` stores info, which stack is currently active (class or method) in the `pointer stack`.
+You can change between class and method resources stack with method `ResourceManager.setMethodResources()` or `ResourceManager.setClassResources()`.
+Note that pointer stack is set automatically in `AbstractST.class` in `@BeforeAll` or `@BeforeEach` methods.
+
+Cluster Operator setup example:
+
+```
+@BeforeAll
+void createClassResources() {
+   ResourceManager.setClassResources();     <--- point resource manager to class stack
+   installClusterOperator(NAMESPACE);       <--- install cluster operator
+}
+```
+
+### Exercise
+In this phase we prepare our component, which is tested. In our case it is set of operation that can modified state of our 
+components for instance updating configuration of `Kafka`, `KafkaConnect` or creating new listeners in `Kafka` and more.
+
+### Verify
+
+When your environment is ready from the previous phase then incomes checks of state, message exchange and so on.
+In this code snippet you can see verification phase, where we verify that configuration of the Kafka custom resource is successfully set.
+Moreover, we verify that specific property is correctly set inside Kafka pod.
+
+```
```

Comment:
Does it makes sense to add the ```java for the all other code blocks?

Reply:
I have tried to have all java code blocks but in our cases, on website it is not pretty.

Reply:
Ah okay, well you have there one java code block, so in case it's not pretty, can you please delete it?